nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,155 @@
+            this.messageContext = messageContext;
+        }
+
+        @Override
+        public void from(String from) throws RejectException {
+            // TODO: possibly whitelist senders?
+            this.from = from;
+        }
+
+        @Override
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(true);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            final long serverTimeout = TimeUnit.MILLISECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+
+            int rd;
+
+            try {
+                while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1 ) {
+                    baos.write(buffer, 0, rd);
+                    if (baos.getBufferLength() > server.getMaxMessageSize() ) {
+                        throw new TooMuchDataException("Data exceeds the amount allowed.");
+                    }
+                }
+                baos.flush();
+            } catch (IOException e) {
+                throw new DropConnectionException(450, "Unexpected error processing your message. ");
+            }
+
+            this.messageData = baos;
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageData, latch);
+            // Try to queue the message back to the NiFi session
+            try {
+                incomingMessages.offer(message, serverTimeout, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                logger.info("The SMTP processor has just dropped a message due to the queue " +
+                        "being too full, considering increasing the queue size");
+                throw new DropConnectionException(421, "Could not queue the message. Try again ");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            while (!message.getProcessed()) {
```

Comment:
I'm confused what added benefit/functionality this while loop provides. There isn't any retry inside and it is just waiting inside the loop anyway. Why not just simplify it to something like:

> try {
>                         latch.await(servertimeout - elapsed, TimeUnit.MILLISECONDS);
>                     } catch (InterruptedException e) {
>                         if (message.getProcessed() == true) {
>                             logger.info("Latch opened unexpectedly but processor indicates data processed, proceeding normally.");
>                         } else {
>                             logger.info("Latch opened unexpectedly and processor indicates data wasn't processed. Returned error to SMTP client as precautionary measure");
>                             throw new DropConnectionException(423, "Unexpected Error. Please try again or contact the administrator in case it persists");
>                         }
>                     }
>  if (!message.getProcessed(){
> ....